exercises/practice/bob/.articles/performance/code/main.rs

```diff
@@ -71,20 +74,132 @@ pub fn reply_array(msg: &str) -> &str {
+    QuestionMarkUpperCase,
+    QuestionMarkUndefined,
+}
+
+const FINE_BE_THAT_WAY: &str = "Fine. Be that way!";
+const WHATEVER: &str = "Whatever.";
+const SURE: &str = "Sure.";
+const CHILL_OUT: &str = "Whoa, chill out!";
+const CALM_DOWN: &str = "Calm down, I know what I'm doing!";
+
+pub fn reply_state_machine(message: &str) -> &str {
+    let message = message.trim();
+    if message.is_empty() {
+        return FINE_BE_THAT_WAY;
+    }
+    let mut state = State::Initial;
+    for c in message.chars().rev() {
+        match state {
+            State::Initial => {
+                state = if c == '?' {
+                    State::HasQuestionMark
+                } else {
+                    if c.is_lowercase() {
+                        return WHATEVER;
+                    }
+                    if c.is_uppercase() {
+                        State::NoQuestionMarkUpperCase
+                    } else {
+                        State::NoQuestionMarkUndefined
+                    }
+                };
+            }
+            State::HasQuestionMark => {
+                state = if c.is_uppercase() {
+                    State::QuestionMarkUpperCase
+                } else {
+                    State::QuestionMarkUndefined
+                };
+            }
+            State::NoQuestionMarkUpperCase => {
+                if c.is_lowercase() {
+                    return WHATEVER;
+                }
+            }
+            State::NoQuestionMarkUndefined => {
+                if c.is_lowercase() {
+                    return WHATEVER;
+                }
+                if c.is_uppercase() {
+                    state = State::NoQuestionMarkUpperCase;
+                }
+            }
+            State::QuestionMarkUpperCase => {
+                if c.is_lowercase() {
+                    return SURE;
+                }
+            }
+            State::QuestionMarkUndefined => {
+                if c.is_lowercase() {
+                    return SURE;
+                }
+                if c.is_uppercase() {
+                    state = State::QuestionMarkUpperCase;
+                }
+            }
+        }
+    }
+    match state {
+        State::HasQuestionMark | State::QuestionMarkUndefined => SURE,
+        State::NoQuestionMarkUpperCase => CHILL_OUT,
+        State::NoQuestionMarkUndefined => WHATEVER,
+        State::QuestionMarkUpperCase => CALM_DOWN,
+        _ => panic!("Unexpected final state"),
+    }
+}
+
 #[bench]
 /// multiple line question for match
 fn multiple_line_question_match(b: &mut Bencher) {
-    b.iter(|| reply_match("\rDoes this cryogenic chamber make me look fat?\rNo."));
+    b.iter(|| {
+        reply_match("\rDoes this cryogenic chamber make me look fat?\rNo.");
+        reply_match("  ");
+        reply_match("Does this cryogenic chamber make me look fat?");
+        reply_match("WHAT'S GOING ON?");
+        reply_match("ZOMG THE %^*@#$(*^ ZOMBIES ARE COMING!!11!!1!");
```

Comment:
A separate function could make sense here, to make it more clear that all benches are running the same code.